Support for interactions and verified bots

doluk opened the pull request:
# Description
Due to the recent changes regarding the permissions of discord bots many verified bots have nowadays only slash commands. Currently those message interactions aren't supported and also there is no difference between a verified bot and a unverified bot. This PR tackles both issues.
### Message Interaction
The implementation is prettly similar to the one for references, just some not needed attributes are removed and the "used /" added. As there is sadly no information in the API available about the arguments, the hover animation is only from cursor to pointer.

### Verified bot batch
The verified symbol is added as svg path to the class. Colors are the same as from the not verified one.

## Former transcript screenshot
![image](https://user-images.githubusercontent.com/69309597/198033399-e98895e7-43d6-4b0c-9e94-6d1969b1220e.png)
## New transcript screenshot
![image](https://user-images.githubusercontent.com/69309597/198033411-2b25e1f3-d087-4d50-b19c-0e52acba2a17.png)
## Actual Discord screenshot
![image](https://user-images.githubusercontent.com/69309597/198033685-eb903cee-d15e-468a-bda9-8425ab2ac72f.png)

mahtoid commented on chat_exporter/construct/message.py:
```diff
@@ -243,6 +271,9 @@ async def generate_message_divider(self, channel_audit=False):
             if self.message.reference != "":
```
` if self.message.reference != "" or self.message.interaction != "":` (check review comment as to why)

mahtoid commented on chat_exporter/construct/message.py:
```diff
@@ -27,7 +29,12 @@
 
 def _gather_user_bot(author: discord.Member):
-    return bot_tag if author.bot else ""
+    if author.bot and author.public_flags and author.public_flags.verified_bot:
```
Is `author.public_flags` necessary? This doesn't return `None` afaik

doluk replied:
I honestly don't know it for sure. So I added it just in case

mahtoid replied:
I'd say return it then, API documentation doesn't say it returns `None`

mahtoid reviewed (changes requested):
There is quite a lot of duplicated CSS between `reference` and `interactions`. It may be good to either:
a) Take on the original CSS from reference
b) Rename the reference CSS to something a bit more common and use it between the two

Additionally, any implementation as you can see around the rest of the project is pretty 1:1 with Discord. There are two things I saw from a quick check:
1. **When referencing a command**
Discord:
![image](https://user-images.githubusercontent.com/56257224/198129451-46b8f28f-1f37-4148-a4fa-efaefaaa74ea.png)
Export:
![image](https://user-images.githubusercontent.com/56257224/198129535-d6111d53-7b3a-4bd8-9c5f-367f06e5fb39.png)
-once you grab the SVG for that, you can make a PR to my [DiscordUtils repo](https://github.com/mahtoid/DiscordUtils)

2. **Underline on hover**
Discord: Displays underline on hover
Export: Does not display underline

---
If you have any question about any of the changes request then as I mentioned on Discord, feel free to ask away and I'll be happy to assist. Either drop a response here, or DM on Discord - either is perfectly fine.

doluk commented on chat_exporter/construct/message.py:
```diff
@@ -27,7 +29,12 @@
+        return bot_tag_verified
+    elif author.bot:
+        return bot_tag
+    else:
```
you mean just return ""? IMO it is more readable wiht the else, but if you want, we can remove it

doluk reviewed (commented):
Regarding the CSS, I would stick with the slightly duplicated one to allow a better customization (in case someone wants to and clearly seperate it).
As I said before, my CSS is not good. I dont know how to make the underline on hover  😅
also for the pr to DiscordUtils, I am not feeling skilled enough to create a properly written pr.
The path for the symbol is `<path fill="#b9bbbe" fill-rule="evenodd" clip-rule="evenodd" d="M5 3C3.89543 3 3 3.89543 3 5V19C3 20.1046 3.89543 21 5 21H19C20.1046 21 21 20.1046 21 19V5C21 3.89543 20.1046 3 19 3H5ZM16.8995 8.41419L15.4853 6.99998L7 15.4853L8.41421 16.8995L16.8995 8.41419Z"></path>`

doluk commented on chat_exporter/html/message/bot-tag-verified.html:
```diff
@@ -0,0 +1,5 @@
+<span class="chatlog__bot-tag-verified">
+    <svg aria-label="Verified Bot" class="chatlog__bot-tag-verified-mark" aria-hidden="false" role="img" width="16"
```
I don't know :D as my CSS/html is bad, I don't know if I have to name it. I just copied the code from discord website and changed the name.

mahtoid replied:
Remove it then, haha. You haven't created a style for it, it's not needed.

doluk commented:
So I tried to remove some css which wasn't used and copied some stuff togehter. Further changes or optimization depends on your decision if you want to use seperate css for interaction and reference or the same.

I added the command icon to appear if a message with a command is referenced, also the text changes to "Click to see command". The svg is a pending pr for your DiscordUtils repo.

I added also the css underline if you hover the command.

I adjusted the parse mdoe for the command name

mahtoid commented on chat_exporter/construct/message.py:
```diff
@@ -146,10 +154,18 @@ async def build_reference(self):
+
 
-        attachment_icon = DiscordUtils.reference_attachment_icon if message.embeds or message.attachments else ""
+        if not message.interaction and (message.embeds or message.attachments):
```
```
icon = ""
if not message.interaction and (message.embeds or message.attachments):
    icon = DiscordUtils.reference_attachment_icon
elif message.interaction:
    icon = DiscordUtils.interaction_command_icon
```

mahtoid commented:
> So I tried to remove some css which wasn't used and copied some stuff togehter. Further changes or optimization depends on your decision if you want to use seperate css for interaction and reference or the same.

Use reference CSS, however if you would like to change the name of the reference CSS to a common name (which would be better, but slightly more work) - then feel free. 

> I added the command icon to appear if a message with a command is referenced, also the text changes to "Click to see command". The svg is a pending pr for your DiscordUtils repo.

https://cdn.jsdelivr.net/gh/mahtoid/DiscordUtils@master/discord-command.svg

doluk commented:
Removed all duplicate css by renaming it to followup. They only split appart in reference and interaction if need.
One small duplicate is reference-icon and interaction-icon but those class names are set in DiscordUtils repo and not here. So unless you wanna change them we have to go with the duplicate.

mahtoid commented on chat_exporter/construct/message.py:
```diff
@@ -45,13 +51,13 @@ class MessageConstruct:
-        military_time: bool,
-        guild: discord.Guild,
-        meta_data: dict
+            self,
```
Indent by 4 not 8 🤔

mahtoid commented on chat_exporter/construct/message.py:
```diff
@@ -67,7 +73,7 @@ def __init__(
 
     async def construct_message(
-        self,
+            self,
```
Indent by 4 not 8 🤔

mahtoid commented on chat_exporter/construct/message.py:
```diff
@@ -223,9 +256,9 @@ async def build_message_template(self):
-            self.previous_message is None or self.message.reference != "" or
-            self.previous_message.author.id != self.message.author.id or self.message.webhook_id is not None or
-            self.message.created_at > (self.previous_message.created_at + timedelta(minutes=4))
+                self.previous_message is None or self.message.reference != "" or self.message.interaction != "" or
```
Indent by 4 not 8 🤔

mahtoid commented on chat_exporter/construct/message.py:
```diff
@@ -340,23 +374,23 @@ def to_local_time_str(self, time):
-    guild: discord.Guild,
-    pytz_timezone,
-    military_time,
+        messages: List[discord.Message],
```
Indent by 4 not 8 🤔

mahtoid commented on chat_exporter/construct/message.py:
```diff
@@ -340,23 +374,23 @@ def to_local_time_str(self, time):
-            military_time,
-            guild,
-            meta_data
+                message,
```
Indent by 4 not 8 🤔

mahtoid commented:
>One small duplicate is reference-icon and interaction-icon but those class names are set in DiscordUtils repo and not here.

Huh? What do you mean

doluk commented:
the name of the element is set in DiscordUtils repo, we just import it here. As the names include interaction and reference, we cant unify it to followup without a change in DiscordUtils

mahtoid commented:
I'm still a little unsure on what you mean. Nothing on `DiscordUtils` repo mentions reference? The only thing I can see in `chat-exporter` is the `discord-attachment` which is set to `reference_attachment_icon` - which is perfectly fine.

doluk commented:
yeah you are right. Idk maybe i wasn't fully awake😅

doluk commented:
i "corrected" the spacing eith the last commit

mahtoid reviewed (approved):
Looks good, tested again and works. May touch at some point in the future, but this is good enough to move in to `development` ahead of release.
